### ISSUE TITLE
sym.Sqrt gradient inf ?

### ISSUE THREAD

**Issue by juliandewit** [dmlc/mxnet#2261]
Hello, I am experimenting a bit with composing layers..
Now my network is a bit unstable. Using the monitor I get "inf" for the gradient of a sqrt symbol.

Now I look at the gradient pass of the sqrt symbol and I see:
`
struct square_root_grad {
  template<typename DType>
  MSHADOW_XINLINE static DType Map(DType a) {
    return DType(DType(0.5f) / a);
  }
};
`

Now I'm not 100% sure of how everything works but doesn't this pose a problem if incoming gradient (a) == 0
?

It would explain my issues..

**Comment by juliandewit** [comment]
Hmmm it's even more strange.
I set minibatch size to 1

The gradient comes from a reshape symbol. Shape is (1,1)
So there is only one value which is 0.49....
As I understand this flows into the sqrt and then the gradient (also shape (1,1) ) is 99.931.
	
INFO:root:Batch:       1 reshape3_backward_data         0.499396	
INFO:root:Batch:       1 sqrt1_backward_data            99.931	

Am I not understanding something here ?
I would expect 0.5 / 0.499 ~ 1.

Doing the reshape before the sqrt gave same result.

**Comment by sxjscience** [comment]
@juliandewit Could you provide a code example? In that case I can replicate the error and debug locally.

**Comment by juliandewit** [comment]
The project as-is is too complex but I'll try to come up with something more isolated.

**Comment by sxjscience** [comment]
Great!

**Comment by juliandewit** [comment]
I have an example that gives another problem but that might explain my problems..
It most probably does not work but I look at the 1st monitor output and it confuses me.

Minibatchsize = 1 so all shapes are (1,1)
Y = 0.071667
output = almost zero
square and sqrt output look correct.
Makeloss output is '0' which seems strange to me but could be...
Makeloss gradient is '1' which also seems strange but could be.
However now sqrt gradient is '6.974...' which I just do not understand.

Am I doing something wrong ? Am I using the software incorrectly ?


`__author__ = 'Julian'
import mxnet as mx
import numpy as np
import logging

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)
np.random.seed(1301)
MINI_BATCH_SIZE = 1
trX = np.linspace(-1, 1, 101)
trY = 2 * trX + np.random.randn(*trX.shape) * 0.33
X = mx.sym.Variable('data')
Y = mx.sym.Variable('softmax_label')
Y = mx.sym.Reshape(Y, shape=(MINI_BATCH_SIZE, 1)) # force right shape to be used later
Y_ = mx.sym.FullyConnected(data=X, num_hidden=1)
diff = mx.sym.square(Y - Y_)
diff = mx.sym.sqrt(diff)
# diff = mx.sym.Reshape(diff, shape=(MINI_BATCH_SIZE, 1))
loss = mx.sym.MakeLoss(diff)
model = mx.model.FeedForward(symbol=loss, num_epoch=1, learning_rate=0.05, numpy_batch_size=MINI_BATCH_SIZE)
mon = mx.mon.Monitor(1)
# mon = None
model.fit(X=trX, y=trY, monitor=mon)`

Now the 1st monitor output is: 
`INFO:root:Batch:       1 reshape0_output                0.0716676	
INFO:root:Batch:       1 fullyconnected0_output         1.95254e-05	
INFO:root:Batch:       1 _minus0_output                 0.0716871	
INFO:root:Batch:       1 square0_output                 0.00513904	
INFO:root:Batch:       1 sqrt0_output                   0.0716871	
INFO:root:Batch:       1 makeloss0_output               0.0	
INFO:root:Batch:       1 makeloss0_backward_data        1.0	
INFO:root:Batch:       1 sqrt0_backward_data            6.97475	
INFO:root:Batch:       1 square0_backward_data          1.0	
INFO:root:Batch:       1 _minus0_backward_lhs           0.0716871	
INFO:root:Batch:       1 _minus0_backward_rhs           1.0	
INFO:root:Batch:       1 fullyconnected0_backward_data  1.0	
INFO:root:Batch:       1 fullyconnected0_backward_weight 0.02	
INFO:root:Batch:       1 fullyconnected0_backward_bias  1.0	
INFO:root:Batch:       1 softmax_label                  0.0716676	
INFO:root:Batch:       1 data                           0.02	
INFO:root:Batch:       1 fullyconnected0_weight         2.373e-05	
INFO:root:Batch:       1 fullyconnected0_bias           0.05`

**Comment by sxjscience** [comment]
I've checked the code. 0.0716871 here means the loss function, which is the output of a sqrt Op. The gradient is thus 0.5/0.0716871 = 6.974.

Also, I've found a bug in makeloss op in https://github.com/dmlc/mxnet/blob/master/src/operator/make_loss-inl.h#L55 that causes Makeloss output to be zero. Since directly using A=B will not copy the value (https://github.com/dmlc/mshadow/issues/50), we need to use `1.0f * data` . I'll make a PR for this.

**Comment by piiswrong** [comment]
The gradient of sqrt at 0 is inf so it's not a stable loss function.
Use sqrt(1+relu(x)) to get a stable loss

**Comment by juliandewit** [comment]
Ok conclusion.
Gradient of SQRT is working as expected.
Only when incoming gradient == 0 it's not stable..
So I will have to stabilize.
Closed.
